tests/test_sdk.py

```diff
@@ -279,6 +282,106 @@ def test_job(hd_client):
+        # look at decorator
+        # verify run start/stop is sent
+        with patch("sys.stdout", new=StringIO()) as faked_out:
+            exp = Experiment("MNIST")
+            exp.log("test print")
+            exp.param("batch size", 32)
+            for i in exp.iter(2):
+                exp.metric("accuracy", i*0.2)
+            time.sleep(0.1)
+            exp.end()
+        
+        # Test params match what is expected
+        params_messages = []
+        for msg in server_sdk_messages:
+            payload = msg["payload"]
+            if "params" in payload:
+                params_messages.append(payload)
+
+        expect_params = [
+            {
+                "params": {
+                    "batch size": 32,
+                },
+                "is_internal": False,
+            },
+            {
+                "params": {
+                    "hd_iter_0_epochs": 2,
+                },
+                "is_internal": True,
+            },
+        ]    
+        assert len(expect_params) == len(params_messages)
+        for i, message in enumerate(params_messages):
+            assert message == expect_params[i]
+
+        # Test metrics match what is expected
+        metrics_messages = []
+        for msg in server_sdk_messages:
+            payload = msg["payload"]
+            if "name" in payload:
+                metrics_messages.append(payload)
+
+        expect_metrics = [
+            {"is_internal": True, "name": "hd_iter_0", "value": 0},
+            {"is_internal": False, "name": "accuracy", "value": 0},
+            {"is_internal": True, "name": "hd_iter_0", "value": 1},
+            {"is_internal": False, "name": "accuracy", "value": 0.2},
+       ]
+        # print(len(expect_metrics), len(metrics_messages))
+        print(expect_metrics)
+        print(metrics_messages)
+        assert len(expect_metrics) == len(metrics_messages)
+        for i, message in enumerate(metrics_messages):
+            # print(expect_metrics[i])
+            assert message == expect_metrics[i]
+        
+        captured_out = faked_out.getvalue()
+        assert "error" not in captured_out
+        
+        # Make sure logs were persisted
+        expect_logs = [
+            "{ batch size: 32 }",
+            "test print",
+            "| Iteration 0 of 1 |",
+            "| accuracy:   0.000000 |",
+        ]
+
+        log_dir = get_hyperdash_logs_home_path_for_job("MNIST")
+        latest_log_file = max([
+            os.path.join(log_dir, filename) for
+            filename in
+            os.listdir(log_dir)
+        ], key=os.path.getmtime)
+        with open(latest_log_file, "r") as log_file:
+            data = log_file.read()
+            for log in expect_logs:
+                assert_in(log, data)
+        os.remove(latest_log_file)
+        
+    def experiment_raises_exceptions(self):
+        exception_raised = True
+        expected_exception = "some_exception_b"
+
+        def test_job():
```

Comment:
did you try doing this manually? Like literally putting this in a python file and seeing what happens? My only fear is that the user's program wont shutdown because your threads wont clean up. I would just verify that the program exits with the exception and doesn't have some zombie bullshit going on

Reply:
Your fears were well founded. The exception cancelled the rest of the function, but the zombie thread lived on. I could kill the thread manually, which would lead to a disconnect, but we obviously can't expect users to do that.

We need better thread cleanup. I'll look into it.

Reply:
Daemonizing the experiment thread fixed it.